programs/mango-v4/src/instructions/perp_liq_negative_pnl_or_bankruptcy.rs

```diff
@@ -129,46 +283,76 @@ pub fn perp_liq_negative_pnl_or_bankruptcy(
             liqee_health_cache.adjust_token_balance(&settle_bank, -settlement)?;
 
             emit!(PerpLiqNegativePnlOrBankruptcyLog {
-                mango_group,
-                liqee: ctx.accounts.liqee.key(),
-                liqor: ctx.accounts.liqor.key(),
+                mango_group: group_key,
+                liqee: liqee_key,
+                liqor: liqor_key,
                 perp_market_index,
                 settlement: settlement.to_bits(),
             });
 
+            liqee_pnl += settlement;
             msg!("liquidated pnl = {}", settlement);
         }
     };
     let max_liab_transfer = I80F48::from(max_liab_transfer) - settlement;
 
-    //
     // Step 2: bankruptcy
     //
+    // If the liqee still has negative pnl and couldn't possibly be settled further, allow bankruptcy
+    // to reduce the negative pnl.
+    //
     // Remaining pnl that brings the account into negative init health is either:
     // - taken by the liqor in exchange for spot from the insurance fund, or
-    // - wiped away and socialized among all perp participants
+    // - wiped away and socialized among all perp participants (this does not involve the liqor)
     //
-    let insurance_transfer = if settlement == max_settlement_liqee {
+    let insurance_transfer;
+    if settlement == max_settlement_liqee && liqee_pnl < 0 {
+        // Preparation that's needed for both, insurance fund based pnl takeover and socialized loss
+
+        let liqee_settle_token_balance = liqee_health_token_balances
+            [liqee_health_cache.token_info_index(settle_token_index)?]
+        .spot_and_perp;
         let liqee_perp_position = liqee.perp_position_mut(perp_market_index)?;
-        let liqee_pnl = liqee_perp_position.unsettled_pnl(&perp_market, oracle_price)?;
 
-        let max_liab_transfer_from_liqee =
-            (-liqee_pnl).min(-liqee_liq_end_health).max(I80F48::ZERO);
-        let liab_transfer = max_liab_transfer_from_liqee
+        // recompute for safety
+        liqee_pnl = liqee_perp_position.unsettled_pnl(&perp_market, perp_oracle_price)?;
+
+        // Each unit of pnl increase (towards 0) increases health, but the amount depends on whether
+        // the health token position is negative or positive.
+        // Compute how much pnl would need to be increased to reach liq end health 0 (while ignoring
+        // liqee_pnl and other constraints for now)
```

Comment:
can the constraints be freely ignored?

Reply:
they are just ignored for max_for_health. The two lines below then limit to liqee_pnl and max_liab_transfer